qiskit_experiments/framework/experiment_data.py

```diff
@@ -938,25 +959,33 @@ def _add_result_data(self, result: Result) -> None:
                 if hasattr(expr_result, "meas_return"):
                     data["meas_return"] = expr_result.meas_return
                 self._result_data.append(data)
+        self.end_datetime = datetime.now()
 
     def _retrieve_data(self):
         """Retrieve job data if missing experiment data."""
-        if self._result_data or not self._backend:
-            return
+        # if self._result_data or not self._backend:
+        #     return
         # Get job results if missing experiment data.
         retrieved_jobs = {}
+        jobs_to_retreive = []
+        if self.job_ids is not None:
+            for jid in self.job_ids:
+                if jid not in self._jobs or self._jobs[jid] is None:
+                    jobs_to_retreive.append(jid)
         for jid, job in self._jobs.items():
             if job is None:
-                try:
-                    LOG.debug("Retrieving job from backend %s [Job ID: %s]", self._backend, jid)
-                    job = self._backend.retrieve_job(jid)
-                    retrieved_jobs[jid] = job
-                except Exception:  # pylint: disable=broad-except
-                    LOG.warning(
-                        "Unable to retrieve data from job on backend %s [Job ID: %s]",
-                        self._backend,
-                        jid,
-                    )
+                jobs_to_retreive.append(jid)
```

Comment:
If `jid` is in `self.job_ids` then you append `jid` twice: here and in line 974.

Reply:
Please document what's the meaning of `self.job_ids, self._jobs, jobs_to_retriece`. As a reader of the code, I'm not following.

Reply:
I agree, we don't need the `for jid, job in self._jobs.items():` part; it's legacy code, and every job id here should already be in `job_ids`.